hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/InodeTree.java

```diff
@@ -133,6 +138,11 @@ public INode(String pathToNode, UserGroupInformation aUgi) {
     // and is read only.
     abstract boolean isInternalDir();
 
+    /**
+     * INode representing a INodeDir which also contains a INodeLink(nested mount point)
+     */
+    abstract boolean isDirAndLink();
+
     // INode linking to another filesystem. Represented
     // via mount table link config entries.
     boolean isLink() {
```

Comment:
This feels wrong. In particular, I'd say that isLink should return true on all links (internal & leaf). This shows as you almost always call x.isDirAndLink() || x.isLink(). Note that the definition would need to change from the current !isInternalDir(). Obviously isInternalDir can be used to determine if it is an internal node versus a leaf.

Reply:
In particular, I'd say that isLink should return true on all links (internal & leaf)
> Yeah, I've considered this but this involves refactoring in traversal/resolve logic and also change the semantics of isLink() which returns true for an internal node. (I am more concerned of this from traversal perspective but it might be contradictory from definition perspective) So I decided to create a new prop to identify internal node with mount point. But the downside is I have to add additional check whenever x.isLink() is invoked.

Reply:
Basically, it looks like half of the 6 callers to isLink should change to !isInternalDir.

Reply:
Changed.